### ISSUE TITLE
Don't create default response or parameter

### ISSUE THREAD

**Issue by michbeck100** [smallrye/smallrye-open-api#24]
When there is no `@Parameter` or `@APIResponse`found on a JAXRS resource method, the `OpenApiAnnotationScanner` creates default values. I think this behavior is not correct. 

1. For the contract-first approach, the parameters and responses might already be described in the static file. At the moment of annotation scanning there is no info about that. 
2. The info might me omitted on purpose. I think the scanner shouldn't assume anything, just because the expected annotations are missing, even if it wasn't on purpose. 

Any thoughts?

**Comment by michbeck100** [comment]
Here is an example for the generated response:

```java
    @POST
    public Response create(Customer customer) {
         // create customer
         return Response.status(Status.CREATED)
                .location(location)
                .tag(tag)
                .entity(createdCustomer)
                .build();
```
The corresponding openapi.yaml looks like this:
```yaml
openapi: 3.0.1
paths:
  /customers:
    post:
      requestBody:
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/Customer'
        description: new customer
        required: true
      responses:
        201:
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Customer'
          description: Customer created
```
When this gets merged the generated openapi document contains both the 201 response and the auto generated 200 response, which is wrong.

**Comment by EricWittmann** [comment]
This is an interesting use case.  What about if the method returned a `Customer` rather than a `Response`.  What then?  I'm not arguing against, necessarily.  Just thinking. :)

**Comment by michbeck100** [comment]
For everything else than `Void` return type, the default response will be generated. see [here](https://github.com/smallrye/smallrye-open-api/blob/4d82fef7fb4f794eafdbe10c87f44a8981776ece/implementation/src/main/java/io/smallrye/openapi/runtime/scanner/OpenApiAnnotationScanner.java#L707). 
For `Void` return types a default response with code 204 or 102 (if POST method) gets generated.

**Comment by EricWittmann** [comment]
Sorry I meant what do you think **should** happen for the various return types?  One of the goals of MP+OAI is to generate as much as possible from a standard JAX-RS application.  Without requiring the developer to necessarily add more annotations.  I think what you've pointed out here is that being aggressive with the generated output can produce conflicts when a static file is also provided.

So I'm wondering what your thoughts are when, for example, a method returns a java bean rather than a `Response`. 

More broadly, perhaps we should investigate the process of merging the OAI documents - adding some additional intelligence to it to resolve these sorts of conflicts.

**Comment by michbeck100** [comment]
> More broadly, perhaps we should investigate the process of merging the OAI documents - adding some additional intelligence to it to resolve these sorts of conflicts.

I agree. The main problem is that the annotation model takes precendence over the static model. If the static model already contains values that are created by the annotation scanner because the annotation is missing, then this is a problem. 

My feeling is that the annotation scanner should do what its name suggests: scan annotations. 
All the additional values should be generated **after** the merge of the static file model and annotations model is completed (and also after filtering).  If there is something missing after all that, then these missing parts can be generated.

**Comment by EricWittmann** [comment]
I understand what you're saying, but it might be a fairly significant change to how this all works.  However, we do need to operate within the constraints of the specification.  Your suggestion would likely violate this section of the spec:

https://github.com/eclipse/microprofile-open-api/blob/master/spec/src/main/asciidoc/microprofile-openapi-spec.adoc#45-processing-rules

I suggested improving the merge logic because the spec is silent on exactly how the various documents should be merged - it only lists the **order** in which processing happens.  So there may be some opportunity there for improvement.

However, when considering this we also need to realize that there are other use-cases that the merging of documents addresses - in fact the spec is designed to solve.  Specifically, the spec calls out the concept of starting from a pure JAX-RS application, generating an OpenAPI document, then modifying that generated document and bundling it back up into the application.  This use-case allows users to e.g. add documentation to their OAI document without adding annotations to the code.

So I guess as I type this, I'm worried that the static + annotation generated use-case (merging of docs) is not meant to solve the case where the JAX-RS annotation processing phase produces something that the user does not want (e.g. Responses, data types, etc).  Instead, developers should customize that output either via a filter (if possible) or by adding OpenAPI specific annotations to the code.

That said, I'm certainly open to changes that improve the utility without violating the specification!!

**Comment by michbeck100** [comment]
The spec says
> If annotation scanning was not disabled, the JAX-RS and OpenAPI annotations from the application will be processed, further overriding any conflicting elements from the current model.

Or in other words: just process the annotations and don't create anything if annotations are missing. I think at least the logic to create an `@APIResponse` from the methods return type should be removed.

**Comment by EricWittmann** [comment]
I think that's not the intention, however.  This section of the spec is perhaps more relevant:

> The MP OpenAPI specification requires vendors to produce a valid OpenAPI document from pure JAX-RS 2.0 applications. 

Along with this, also from the spec:

> ...vendors must process all the relevant JAX-RS annotations (such as @Path and @Consumes) as well as Java objects (POJOs) used as input or output to JAX-RS operations...

This is perhaps an issue we can raise with the specification to see what the intent is.  But I **believe** the intent is to produce a valid OpenAPI document from the available JAX-RS (and plain java) information, which includes processing inputs and outputs as best we can.  If we don't create a response for operations without explicit annotations, the generated OAI doc will be invalid.

**Again, I'd be happy to raise this issue with the spec (or better yet, you can do so)!**

But I do believe that the intent of the spec is to support either code-first or design-first use cases.  Code-first is supported via processing of annotations (JAX-RS, java, OAI annotations).  Design first is supported by including a static file and then disabling annotation scanning.  Your use-case is a bit of a hybrid, where you are saying "I want **some** of the information to come from the source and **some** to come from the static file.  And while the spec seems to support that use-case, I think it was meant more for things like extended documentation, not functional elements.

**Comment by MikeEdgar** [comment]
I think the functionality requested here was effectively implemented in #152, as discussed in #100  (i.e. suppress generating defaults when empty `@APIResponses()` is present). If there are no objections, I'd like to close this issue.

**Comment by MikeEdgar** [comment]
Closing issue. This can be reopened in the future or a new issue opened if new concerns arise.
